File: clients/app/ios/PetSitter/Data/Repositories/AuthRepositoryImpl.swift
import Foundation

// AuthRepositoryImpl: AuthRepositoryProtocol 구현체
// DTO ↔ Domain 변환 책임 — ViewModel은 Domain 모델만 사용
// @unchecked Sendable: 내부 상태 없음, 모든 메서드는 APIClient를 통해 비동기 호출
final class AuthRepositoryImpl: AuthRepositoryProtocol, @unchecked Sendable {
    static let shared = AuthRepositoryImpl()
    private init() {}

    func login(email: String, password: String) async throws -> (accessToken: String, refreshToken: String, user: User) {
        let res: LoginResponseDTO = try await APIClient.shared.requestPublic(
            "/sessions", method: "POST",
            body: LoginRequestDTO(email: email, password: password)
        )
        let me = try await getMe()
        return (res.accessToken, res.refreshToken, me)
    }

    func register(email: String, password: String, name: String, role: String) async throws -> User {
        let dto: UserDTO = try await APIClient.shared.requestPublic(
            "/users", method: "POST",
            body: RegisterRequestDTO(email: email, password: password, name: name, role: role)
        )
        return dto.toDomain()
    }

    func getMe() async throws -> User {
        let dto: UserDTO = try await APIClient.shared.request("/users/me")
        return dto.toDomain()
    }

    func logout() async throws {
        let _: EmptyResponse = try await APIClient.shared.request("/sessions", method: "DELETE")
    }
}

// DELETE /sessions 처럼 응답 바디가 없는 경우 처리
private struct EmptyResponse: Decodable {}

File: clients/app/android/app/src/main/java/com/petsitter/data/repository/AuthRepositoryImpl.kt
package com.petsitter.data.repository

import com.petsitter.data.remote.api.AuthApi
import com.petsitter.data.remote.dto.LoginRequestDto
import com.petsitter.data.remote.dto.RegisterRequestDto
import com.petsitter.data.remote.dto.UserDto
import com.petsitter.domain.model.User
import com.petsitter.domain.model.UserRole
import com.petsitter.domain.repository.AuthRepository
import javax.inject.Inject
import javax.inject.Named
import retrofit2.Retrofit

class AuthRepositoryImpl @Inject constructor(
    private val authApi: AuthApi,
    @Named("private") private val privateRetrofit: Retrofit,
) : AuthRepository {

    // 로그인 후 토큰 반환 — 토큰 저장은 ViewModel에서 처리 (Repository는 저장 책임 없음)
    override suspend fun login(email: String, password: String): Pair<String, String> {
        val res = authApi.login(LoginRequestDto(email, password))
        return Pair(res.accessToken, res.refreshToken)
    }

    override suspend fun register(
        email: String,
        password: String,
        name: String,
        role: String,
    ): User {
        val dto = authApi.register(RegisterRequestDto(email, password, name, role))
        return dto.toDomain()
    }

    override suspend fun getMe(): User {
        // getMe는 인증이 필요하므로 privateRetrofit 사용
        val api = privateRetrofit.create(AuthApi::class.java)
        return api.getMe().toDomain()
    }

    override suspend fun logout() {
        val api = privateRetrofit.create(AuthApi::class.java)
        api.logout()
    }
}

private fun UserDto.toDomain() = User(
    id = id,
    email = email,
    name = name ?: "",
    role = when (role) {
        "PetOwner"  -> UserRole.PetOwner
        "PetSitter" -> UserRole.PetSitter
        else        -> UserRole.Admin
    },
    profileImage = profileImage,
)
